source/val/validate_image.cpp

```diff
@@ -2130,6 +2151,58 @@ spv_result_t ValidateImageSparseTexelsResident(ValidationState_t& _,
   return SPV_SUCCESS;
 }
 
+spv_result_t ValidateImageProcessingQCOMDecoration(ValidationState_t& _, int id,
+                                                   spv::Decoration decor) {
+  const Instruction* si_inst = nullptr;
+  const Instruction* ld_inst = _.FindDef(id);
+  if (ld_inst->opcode() == spv::Op::OpSampledImage) {
+    si_inst = ld_inst;
+    const spv_parsed_operand_t& topnd = si_inst->operand(2);  // texture
+    int t_idx = (int)si_inst->word(topnd.offset);
+    ld_inst = _.FindDef(t_idx);
+  }
+  assert(ld_inst->opcode() == spv::Op::OpLoad);
+  const spv_parsed_operand_t& popnd = ld_inst->operand(2);  // variable to load
+  int texture_id = (int)ld_inst->word(popnd.offset);
+  if (!_.HasDecoration(texture_id, decor)) {
+    return _.diag(SPV_ERROR_INVALID_DATA, ld_inst) << "Missing decoration";
+  }
+
+  return SPV_SUCCESS;
+}
+
+spv_result_t ValidateImageProcessingQCOM(ValidationState_t& _,
+                                         const Instruction* inst) {
+  spv_result_t res = SPV_SUCCESS;
+  const spv::Op opcode = inst->opcode();
+  switch (opcode) {
+    case spv::Op::OpImageSampleWeightedQCOM: {
+      const spv_parsed_operand_t& wopnd = inst->operand(4);  // weight
+      int wi_idx = (int)inst->word(wopnd.offset);
+      res = ValidateImageProcessingQCOMDecoration(
+          _, wi_idx, spv::Decoration::WeightTextureQCOM);
+      break;
+    }
+    case spv::Op::OpImageBlockMatchSSDQCOM:
+    case spv::Op::OpImageBlockMatchSADQCOM: {
+      const spv_parsed_operand_t& topnd = inst->operand(2);  // target
+      int tgt_idx = (int)inst->word(topnd.offset);
+      res = ValidateImageProcessingQCOMDecoration(
+          _, tgt_idx, spv::Decoration::BlockMatchTextureQCOM);
+      const spv_parsed_operand_t& ropnd = inst->operand(4);  // reference
+      int ref_idx = (int)inst->word(ropnd.offset);
+      spv_result_t res1 = ValidateImageProcessingQCOMDecoration(
+          _, ref_idx, spv::Decoration::BlockMatchTextureQCOM);
+      res = (res != SPV_SUCCESS ? res : res1);
```

Comment:
Why not just return earlier if `res` is not success?

Reply:
Done.